ethers-etherscan/src/contract.rs

```diff
@@ -1,141 +1,240 @@
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, path::Path};
 
-/// Arguments for verifying contracts
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct VerifyContract {
-    #[serde(rename = "contractaddress")]
-    pub address: Address,
-    #[serde(rename = "sourceCode")]
-    pub source: String,
-    #[serde(rename = "codeformat")]
-    pub code_format: CodeFormat,
-    /// if codeformat=solidity-standard-json-input, then expected as
-    /// `erc20.sol:erc20`
-    #[serde(rename = "contractname")]
-    pub contract_name: String,
-    #[serde(rename = "compilerversion")]
-    pub compiler_version: String,
-    /// applicable when codeformat=solidity-single-file
-    #[serde(rename = "optimizationUsed", skip_serializing_if = "Option::is_none")]
-    pub optimization_used: Option<String>,
-    /// applicable when codeformat=solidity-single-file
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub runs: Option<String>,
-    /// NOTE: there is a typo in the etherscan API `constructorArguements`
-    #[serde(rename = "constructorArguements", skip_serializing_if = "Option::is_none")]
-    pub constructor_arguments: Option<String>,
-    /// applicable when codeformat=solidity-single-file
-    #[serde(rename = "evmversion", skip_serializing_if = "Option::is_none")]
-    pub evm_version: Option<String>,
-    #[serde(flatten)]
-    pub other: HashMap<String, String>,
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub enum SourceCodeLanguage {
+    #[default]
+    Solidity,
+    Vyper,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SourceCodeEntry {
+    pub content: String,
+}
+
+impl<T: Into<String>> From<T> for SourceCodeEntry {
+    fn from(s: T) -> Self {
+        Self { content: s.into() }
+    }
 }
 
-impl VerifyContract {
-    pub fn new(
-        address: Address,
-        contract_name: String,
-        source: String,
-        compiler_version: String,
-    ) -> Self {
-        Self {
-            address,
-            source,
-            code_format: Default::default(),
-            contract_name,
-            compiler_version,
-            optimization_used: None,
-            runs: None,
-            constructor_arguments: None,
-            evm_version: None,
-            other: Default::default(),
+/// The contract metadata's SourceCode field.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum SourceCodeMetadata {
+    /// Contains metadata and path mapped source code.
+    Metadata {
+        /// Programming language of the sources.
+        #[serde(default, skip_serializing_if = "Option::is_none")]
+        language: Option<SourceCodeLanguage>,
+        /// Source path => source code
+        #[serde(default)]
+        sources: HashMap<String, SourceCodeEntry>,
+        /// Compiler settings, None if the language is not Solidity.
+        #[serde(default, skip_serializing_if = "Option::is_none")]
+        settings: Option<Settings>,
```

Comment:
I wonder if we should just keep this as serde_json::Value and feature get the `fn settings -> Settings` function so we can make ethers-sol optional